File: LLM_experiments/test_results/test_solution_version_GoalInstructionOutput_v1/src/alignment/needleman_wunsch.py
from typing import Dict

def calculate_alignment_score(seq1: str, seq2: str, scoring_matrix: Dict[str, int]) -> int:
    """
    Calculates the optimal global alignment score using the Needleman-Wunsch algorithm.

    Args:
        seq1: The first sequence (protein).
        seq2: The second sequence (protein).
        scoring_matrix: A dictionary containing substitution scores (e.g., "AA")
                        and character-dependent gap penalties (e.g., "A").

    Returns:
        The final integer alignment score.
    """
    n, m = len(seq1), len(seq2)
    dp_matrix = [[0] * (m + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        dp_matrix[i][0] = dp_matrix[i - 1][0] + scoring_matrix[seq1[i - 1]]
    for j in range(1, m + 1):
        dp_matrix[0][j] = dp_matrix[0][j - 1] + scoring_matrix[seq2[j - 1]]

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            char1, char2 = seq1[i - 1], seq2[j - 1]
            match_key, rev_match_key = char1 + char2, char2 + char1
            match_score = scoring_matrix.get(match_key, scoring_matrix.get(rev_match_key))
            
            if match_score is None:
                raise KeyError(f"Scoring for pair '{match_key}' or '{rev_match_key}' not found in matrix.")

            score_diag = dp_matrix[i - 1][j - 1] + match_score
            score_up = dp_matrix[i - 1][j] + scoring_matrix[char1]
            score_left = dp_matrix[i][j - 1] + scoring_matrix[char2]

            dp_matrix[i][j] = max(score_diag, score_up, score_left)

    return dp_matrix[n][m]

File: LLM_experiments/java/results_claude-opus-4/solution_try_1/src/main/java/com/phylogenetic/alignment/NeedlemanWunsch.java
package com.phylogenetic.alignment;

import com.phylogenetic.model.BlosumMatrix;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Implementation of the Needleman-Wunsch global sequence alignment algorithm.
 */
public class NeedlemanWunsch {
  private static final Logger logger = LoggerFactory.getLogger(NeedlemanWunsch.class);
  private final BlosumMatrix blosumMatrix;
  
  public NeedlemanWunsch(BlosumMatrix blosumMatrix) {
    this.blosumMatrix = blosumMatrix;
  }
  
  /**
   * Calculates the alignment score between two sequences.
   *
   * @param sequence1 first amino acid sequence
   * @param sequence2 second amino acid sequence
   * @return the alignment score
   */
  public int calculateScore(String sequence1, String sequence2) {
    int m = sequence1.length();
    int n = sequence2.length();
    
    // Initialize the scoring matrix
    int[][] scoreMatrix = new int[m + 1][n + 1];
    
    // Initialize first row with gap penalties
    for (int j = 1; j <= n; j++) {
      scoreMatrix[0][j] = scoreMatrix[0][j - 1] 
          + blosumMatrix.getGapPenalty(sequence2.charAt(j - 1));
    }
    
    // Initialize first column with gap penalties
    for (int i = 1; i <= m; i++) {
      scoreMatrix[i][0] = scoreMatrix[i - 1][0] 
          + blosumMatrix.getGapPenalty(sequence1.charAt(i - 1));
    }
    
    // Fill the scoring matrix
    for (int i = 1; i <= m; i++) {
      for (int j = 1; j <= n; j++) {
        char char1 = sequence1.charAt(i - 1);
        char char2 = sequence2.charAt(j - 1);
        
        // Calculate scores for three possible operations
        int matchScore = scoreMatrix[i - 1][j - 1] 
            + blosumMatrix.getSubstitutionScore(char1, char2);
        int deleteScore = scoreMatrix[i - 1][j] 
            + blosumMatrix.getGapPenalty(char1);
        int insertScore = scoreMatrix[i][j - 1] 
            + blosumMatrix.getGapPenalty(char2);
        
        // Take the maximum score
        scoreMatrix[i][j] = Math.max(matchScore, Math.max(deleteScore, insertScore));
      }
    }
    
    return scoreMatrix[m][n];
  }
}